Added a units select list to the advanced search

hepcat72 opened the pull request:
## Summary Change Description

These changes add a units select list (currently only for the fields that are DurationFields: Animal.age and Sample.time_collected).  The select list only appears where one of those fields is selected.  If it's a complex format, a question mark icon will appear and if you hover over it, an explanation of the input format will appear.  Also, changing the units changes the placeholder test in the search term field with an example.

When a search term is entered and units selected, the search term will either be converted to the native units or add formatting to make the query valid when taking how the field values are stored.

## Affected Issue Numbers

- Resolves #231

## Code Review Notes

Currently, basic search links could be supported, as I updated the function underneath, but I didn't test or check using units in a basic search.

I had to squash the last 12 commits in order to have a clean merge with main.

## Checklist

- [x] All issue requirements satisfied (or no linked issues)
- [x] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
- [x] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
- Tests
  - [x] [Tests implemented *(or no code changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [x] [All *multi_working* tag tests pass locally](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All example load tests pass (remotely) *(or their failures predated and are unrelated to this branch)*
  - [x] `@tag("multi_working")` added to newly passing test functions/classes *(or no newly passing tests)*
  - [x] `@tag("multi_broken")`, `@tag("multi_unknown")`, or `@tag("multi_mixed")` removed from newly passing test functions/classes *(or no newly passing tests)*

hepcat72 commented on DataRepo/formats/dataformat.py:
```diff
@@ -70,9 +70,128 @@ class Format:
             ("isnull", "does not have a value (ie. is None)"),
         ],
     }
```
C. The following are the basic units options.  There are 2 groups of units that go together: `postgres_interval` and `identity`.  Each group has a `default` unit and a dict containing the different units in the set named `entry_options`.  Inside `entry_options` is a dict describing each specific unit that can populate a select list, e.g. years, months, weeks, days,... (these keys are the values in a units select list).  And each unit can have up to 4 values:

- name - the display value in the units select list
- example - What will appear as the placeholder in the search term box
- convert - A function to massage what the user entered into a format needed by the database
- about - If the units format is complex, a description provided here will appear as a tooltip

This structure is the "global" definition of units.  It contains everything.  Derived classes of this class have a means to select a different default and subset of units.

lparsons replied:
This is a flexible method of defining the unit options. If I'm interpreting your notes correctly, we can specify a subset of these for each field, which is nice so we don't overwhelm the user.

hepcat72 replied:
Yes, that's right.

hepcat72 commented on DataRepo/formats/fluxcirc_dataformat.py:
```diff
@@ -156,10 +156,22 @@ class FluxCircFormat(Format):
+                    "searchable": True,
                     "displayed": True,
                     "type": "number",
+                    "units": {  # See dataformat.py: Format.unit_options
+                        "key": "postgres_interval",
+                        "default": "weeks",
+                        "subset": [
+                            "native",
+                            "calendartime",
+                            "months",
+                            "weeks",
+                            "days",
+                            "hours",
+                        ],
+                    },
```
C. This is the first example where for a specific model(instance) field can specify a different default unit and the subset.  The key specifies the units set to get the available units from.

hepcat72 commented:
I bet it treats it like a float, where there's no equality.  Loading my sandbox database currently because my current data is missing ages.  Did you try any other searches (like greater than/less than)?  I tried time_collected > 140 minutes and that worked.  I got records where the time_collected was 150.  Does that work for you?

fkang-pu reviewed (changes requested):
After review the code and ran some tests, I have comments in two categories:

1. test "units" features:

   Tests worked for me, but the display for units was not as I expected for some cases.

   e.g.
   if I search age>=90 days, after clicking "search" button, I got results but the form on top showed as Age >= 90 weeks.  ( I think it should display as 90 days). It happened to other non-default types too.

2. related to design decision:
  It's a feature enhancement. Big-picture wise, I am not sure if we need/want to add units.  The "units" feature works for PostgreSQL database. If we want other outside groups to use it later, the question is if we want the code to be more flexible (e.g. run on other database platform with minor changes). 

   An alternative is to offer less flexible search with age and time_collected  in fixed unit type ( e.g. age (in weeks), time_collected ( minutes after infusion) , similar to Body Weight(g)). I can see the pros and cons, just think out loud.

hepcat72 commented:
> After review the code and ran some tests, I have comments in two categories:
> 
> 1. test "units" features:
>    Tests worked for me, but the display for units was not as I expected for some cases.
>    e.g.
>    if I search age>=90 days, after clicking "search" button, I got results but the form on top showed as Age >= 90 weeks.  ( I think it should display as 90 days). It happened to other non-default types too.

OK, this has to do with the javascript code and it's initial form population.  The results are fine.  It's just the original search wasn't correctly pre-populated.  Should be an easy fix.

- [ ] Debug the initial form population from existing searches.

> 2. related to design decision:
>    It's a feature enhancement. Big-picture wise, I am not sure if we need/want to add units.  The "units" feature works for PostgreSQL database. If we want other outside groups to use it later, the question is if we want the code to be more flexible (e.g. run on other database platform with minor changes).
>    An alternative is to offer less flexible search with age and time_collected  in fixed unit type ( e.g. age (in weeks), time_collected ( minutes after infusion) , similar to Body Weight(g)). I can see the pros and cons, just think out loud.

I have a 2 category response as well.

1. Code review is not an ideal place for design review.  The point of a design review is to solve problems *before* implementation, in order to not waste implementation effort.  We have a design review structure implemented that lays out design implementation in the issues themselves and labels to indicate a review should be done and an approval process for handling those reviews, but that shouldn't happen here in a code review - it should take place in the design embedded in the issues.  I have proposed design review processes in the past, but we as a group, have not adopted them.  The problem is that we don't have a good process for determining when/if those reviews should occur.  We don't require them and we don't track them, so the onus is on individual developers to apply the "needs-review" label if they feel there's an issue or a potential issue associated with planned work in the issues interface.  The result is friction in these code reviews where developers have completely different design philosophies where one developer frequently wants to discard work that another developer (who put in the effort and thoughtful work) feels proud of.  I suggest that since you frequently take design-level issues with my efforts, you should review my issues and raise your design concerns there.
2. This feature is a bug fix as opposed to an enhancement and it is explicitly designed to *disconnect* the database from the base advanced search code^.  Flexibility is what the `convert` and `pyconvert` functions coded into the configuration are for.  And this code is explicitly designed to be applied to any project, let alone other instances of Tracebase.  In fact, the code *before* this PR was tied to the postgres database.  It only worked because we were using postgres.  This units interface now allows the code to work with other database architectures.  All one has to do is create and populate a derived class, which includes creating the convert functions.  In fact, it will work with any ORM field type, not just DurationField.

You requested changes Fan, which implies (given our convention) that one or more of your issues is blocking, but you did not specify which issues you raised are blocking.  So I will assume it is the initial select list population.  Since I think the second issue is not valid, I will leave that component of the PR as-is and fix the initial select list population issue you uncovered.  I'd fixed a similar issue before, but I suspect that perhaps the last fix broke it for the units select list.

*^ - It's true that the current convert functions still take advantage of postgres, but the convert functions are anonymous functions specifically because the derived class in which they reside is essentially a config class, and is meant to be modified.  I suspect that making it database agnostic is for them to output timedelta objects, but I was only focused on separating the base code from the database, not the config.*

hepcat72 commented:
I also think that we should adopt a policy of encouraging *constructive* and *specific* feedback.  I think that destructive feedback (i.e. rejecting work/effort) without discrete references in the code is harmful.  For example, if you feel that this work ties the code to postgres, attach that feeback to the specific piece of code that you think is what makes the code database-specific.  And even then, if you find the specific code that has this flaw, it would be better to suggest how to make it not tied to the database, or at the least, point out how that piece of code makes the code database-specific.

hepcat72 commented:
OK Fan, the issues you raised have been addressed.  I fixed the initial population of the units select list and I have re-requested your review.

fkang-pu commented:
@hepcat72 I hope you don't take my comments personally, since my intention was to share my thoughts about the project.  I agree with you that code review is not a good place for design review. I think the root problem is we often don't have a process for discussing design, prototypes before implementing feature(s). 

I might have missed details or misunderstood something, but I was not aware of the  "units" feature is a bug fixing as [issue 231](https://github.com/Princeton-LSI-ResearchComputing/tracebase/issues/231) is labeled as "type:enhancement".

Sorry I didn't make it clear that the only blocking issue was display of correct unit type with advanced search. I fetched and merged the code you updated yesterday in my local branch. Just tried "Advanced Search" page, somehow the dynamic search form disappeared.  Is it working on your end?

hepcat72 commented:
> @hepcat72 I hope you don't take my comments personally, since my intention was to share my thoughts about the project. I agree with you that code review is not a good place for design review. I think the root problem is we often don't have a process for discussing design, prototypes before implementing feature(s).
> 
> I might have missed details or misunderstood something, but I was not aware of the "units" feature is a bug fixing as [issue 231](https://github.com/Princeton-LSI-ResearchComputing/tracebase/issues/231) is labeled as "type:enhancement".
> 
> Sorry I didn't make it clear that the only blocking issue was display of correct unit type with advanced search. I fetched and merged the code you updated yesterday in my local branch. Just tried "Advanced Search" page, somehow the dynamic search form disappeared. Is it working on your end?

OK, I have fixed the issue.  I realized that the place where I'd handled the "initial value" of submitted forms was not exclusively for populated search forms, but some other case having to do with forms for other formats.  I have some tests to fix due to some changed return values, but those will be trivial to fix.

Regarding the "bug" versus "feature" question, the associated issue is somewhat of a gray area.  The code was originally intended to work with any underlying database, but it turned out to only work with the current postgres database.  In my mind, that was a bug, even though nothing was broken, per se.  If we'd switched databases, it wouldn't work.  There had also been an issue with a prior format description of "d-hh:mm:ss", which had been fixed prior to this PR to be "hh:mm:ss" because that other format didn't actually work, but also, it seemed like a bug to make the user multiply the weeks entry they wanted by 7 and 24 to come up with (for example) 14 weeks -> "2352:00:00".  That was just ridiculous.

The code I implemented works for both animal age (in weeks) and sample time_collected in minutes, and to make the code scalable and flexible, I decided to use the same code for both instead of hard-coding static units for each field.  It is this sort of decision that I feel strongly: reduces downstream maintenance costs.  Even though that code is *initially* simpler to implement, those simple decisions in development multiple and intertwine to create very difficult to maintain code.  It creates unknown side-effects.  Fixes to bugs in that code can cause unexpected other problems and defect frequency rises.  In the end, you often end up with a beast of a codebase that turns out to be more complex than what you would have if you'd gone to the effort to create a generic solution in the first place.  This is something I learned formally, from an industry software engineering expert who moonlighted as a prof on the UNM satellite campus in Los Alamos.  And it wasn't just his opinion.  He presented large amounts of project audit data and performance analyses and assessments.  He was essentially a software engineering evangelist.  My boss and I took the class together and we immediately implemented what we learned in our team development efforts.  And one of the biggest things we learned was where most of the costs are in software development: the maintenance phase.  As a rule of thumb, the more software engineering processes employed, the lower the maintenance costs.  And, the more effort expended in the earliest portions of the software development lifecycle, the lower the maintenance costs.  Bugs introduced in the earliest phases (planning, requirements, design) are more expensive to fix the later they are caught.  This case of having design "bugs" discussed during the implementation and testing phases is a case-in-point.  I've already spent the time (that's the "cost") implementing the code and worked on tests.  If we find a "bug" in the design, that has already been implemented, then the solution is a refactor, which means the work done before is lost/wasted and the implementation cost doubles, if not more than doubles.  Now, I don't think there is a design bug, but you can see what I mean.

It's not necessarily that I'm taking your feedback personally.  It's that I have frustration that I know solutions to the problems that this sort of repeated interaction is the result of, and despite my efforts to solicit buy-in from others on the team to adopt those solutions (e.g. a clear process to govern design reviews, awareness of the value of scalability and robustness to reduce downstream maintenance costs, clearly defined roles and goals during the code review process, etc), I feel it's a lost cause.  And even though I try to hold my own personal efforts on this project to those standards, I feel I am met with increasing resistance to the point where I almost feel like my efforts are seen as a liability as opposed to a benefit.  I'm not the best coder in the world.  Far from it.  I know a good portion of my own coding deficits.  But I'm realizing that without reciprocal investments, I feel I benefit less from the code reviews, (e.g. if others on the team take exception to the design, they don't feel obliged to understand the code enough to provide a detailed critique and instead fall back to holistically call the design decisions into question).  When I put a PR out, I'm looking for help to make the code better, not try and justify that my design decisions were correct.  The principles behind those design decisions are the same ones I've been making all along.  A code review should be detailed and technical, to determine how well it meets the issue requirements, how well it adheres to the design, and if it has any defects based on those requirements and design.  It should be pretty banal in that regard.  Instead, I feel like it repeatedly turns into a debate that has nothing to do with the technical correctness of the code, which is what a typical code-review prescribes.

hepcat72 commented on DataRepo/formats/dataformat.py:
```diff
@@ -782,6 +1162,34 @@ def meetsAllConditionsByValList(self, rootrec, query, field_order):
+        takes the field path, the search term value, and the units, and returns the search term value in the units/
+        format that the database returns so that they can be compared in `self.meetsCondition()`.
+
+        If
```
- [ ] TODO: Fix incomplete docstring

hepcat72 commented on DataRepo/formats/dataformat.py:
```diff
@@ -782,6 +1162,34 @@ def meetsAllConditionsByValList(self, rootrec, query, field_order):
+            if units in units_lookup[fld].keys():
+                try:
+                    return units_lookup[fld][units]["pyconvert"](val)
+                except Exception as e:
```
It returns the `val` that was supplied, which yes - would be incorrect in the context of the selected units.  I had used `Exception` because the convert functions are specified in the config.  They could be anything, and they can be for different field types.

Upon thinking about it, I agree that there can be no entirely graceful outcome here without executing an incorrect search.  The only graceful option would be to inform the user that there's a problem with the code behind the units they selected, and that the issue has been reported to the developers.  It could direct them back to the previous page and suggest they try different units... At least, these would be my druthers.  For now, I will:

- [ ] TODO: Raise an exception that suggests the user try different units.

hepcat72 commented on DataRepo/formats/fluxcirc_dataformat.py:
```diff
@@ -156,10 +156,22 @@ class FluxCircFormat(Format):
+                    "searchable": True,
                     "displayed": True,
                     "type": "number",
+                    "units": {  # See dataformat.py: Format.unit_options
+                        "key": "postgres_interval",
+                        "default": "weeks",
+                        "subset": [
+                            "identity",
+                            "calendartime",
+                            "months",
+                            "weeks",
+                            "days",
+                            "hours",
+                        ],
+                    },
```
I'll have to look at the javascript code.  It might assume `identity` is always present.  I thought about that as I coded it and thought I should make it more dynamic, but I didn't get around to it.

- [ ] Make the units options be "months", "weeks", "days", and "hours".  make sure omitting "identity" won't cause problems with the javascript, which has "identity" hard-coded (if I recall correctly).

hepcat72 commented on DataRepo/formats/fluxcirc_dataformat.py:
```diff
@@ -262,10 +274,21 @@ class FluxCircFormat(Format):
+                        "key": "postgres_interval",
+                        "default": "minutes",
+                        "subset": [
+                            "identity",
+                            "clocktime",
+                            "hours",
+                            "minutes",
+                            "seconds",
```
- [ ] TODO: "hours", "minutes", and "seconds"

hepcat72 commented on DataRepo/formats/peakdata_dataformat.py:
```diff
@@ -360,10 +360,22 @@ class PeakDataFormat(Format):
+                    "searchable": True,
                     "displayed": True,
                     "type": "number",
+                    "units": {  # See dataformat.py: Format.unit_options
+                        "key": "postgres_interval",
+                        "default": "weeks",
+                        "subset": [
+                            "identity",
+                            "calendartime",
+                            "months",
+                            "weeks",
+                            "days",
+                            "hours",
+                        ],
+                    },
```
- [ ] TODO: "months", "weeks", "days", and "hours"

hepcat72 commented on DataRepo/formats/peakgroups_dataformat.py:
```diff
@@ -286,10 +286,22 @@ class PeakGroupsFormat(Format):
+                    "searchable": True,
                     "displayed": True,
                     "type": "number",
+                    "units": {  # See dataformat.py: Format.unit_options
+                        "key": "postgres_interval",
+                        "default": "weeks",
+                        "subset": [
+                            "identity",
+                            "calendartime",
+                            "months",
+                            "weeks",
+                            "days",
+                            "hours",
+                        ],
+                    },
```
- [ ] TODO: "months", "weeks", "days", and "hours"
